pkg/network/common/dns.go

```diff
@@ -90,36 +103,77 @@ func (d *DNS) Delete(dns string) {
 	delete(d.dnsMap, dns)
 }
 
-func (d *DNS) Update(dnsName string) (bool, error) {
+// Update updates the TTL, IPs and nextQueryTime for the dnsName.
+// When calling privately, make sure it isn't called a function that holds
+// DNS.lock.
+func (d *DNS) Update(dns string) (bool, error) {
+	// GetNextQueryTime always returns the dns with the earliest
+	// NextQueryTime. If GetNextQueryTime is called while Update is running it
+	// will return the same dns again, so we need to prevent that from
+	// happening.
+	err := d.blockName(dns)
+	if err != nil {
+		return false, err
+	}
+
+	// This has to be done outside the lock because it may block for a while
+	ips, ttl, err := d.getIPsAndMinTTL(dns)
+
 	d.lock.Lock()
 	defer d.lock.Unlock()
 
-	return d.updateOne(dnsName)
+	if err != nil {
+		res, ok := d.dnsMap[dns]
+		if !ok {
+			// Can happen if DNS.Delete was called in the middle of an operation.
+			// In such cases we want to abort the update operation.
+
+			// In this scenario we don't really need to report the error because
+			// the record is gone anyway.
+			return false, fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
```

Comment:
you say you don't need to report the error, but then you report the error

Reply:
Oh, by the error I meant the`err` variable.

Other than the the comment are you OK with the rest of the changes? If so I'll squash the tmp commit and change the comment for: In this scenario we don't really need to report the getIPsAndMinTTL error because the record is gone anyway.

Reply:
I was hoping for a little more rewriting... if there was some way to not need to grab locks and then release them and then grab them again, then the code would all make a lot more sense. I just feel like if we merge this we're going to be back here again in a few weeks.

(I don't actually see any bugs in the current code, but then, I didn't notice any bugs in the previous code either.)

Reply:
You're making perfect sense.

I added yet another commit which makes things cleaner, not good yet, but they're a step (I think) in the right direction. Right now I removed the `d.blockName` function by setting the `dnsValue.updating` field to true in `d.GetNextQueryTime`, which means now the Update function doesn't need to lock, unlock and lock again.

There is still a problem with the locking, both `Update` and `Add` call `getIPsAndMinTTL` before locking, but I can't think of a way to avoid this and still make the query async.

PS: I'm adding additive commits to make review easier, or do you prefe just amended commits? do you have any preference?

Reply:
/hold
This is just wrong. After GetNextQueryTime is executed there is no guarnatee that Update is going to be called again. this is just wrong.

Reply:
> PS: I'm adding additive commits to make review easier, or do you prefe just amended commits? do you have any preference?

In this particular case I'd rather you just amended commits, because the problem is that the first version of the patch was confusing, so I'd rather compare the new less-confusing version against the original code than against the confusing intermediate version.